File: web/src/api/task.ts
import apiClient from './client'
import type { Workflow } from './workflow'
import type { MediaAsset } from './asset'

export interface Task {
  id: string
  workflow_id: string
  workflow_revision_id?: string
  workflow_revision?: number
  asset_id?: string
  status: 'pending' | 'running' | 'success' | 'failed' | 'cancelled'
  progress: number
  current_node?: string
  context_version?: number
  input_params?: Record<string, any>
  error?: string
  started_at?: string
  completed_at?: string
  created_at: string
  updated_at: string
  workflow?: Workflow
  asset?: MediaAsset
  node_executions?: any[]
}

export interface TaskListQuery {
  workflow_id?: string
  asset_id?: string
  status?: 'pending' | 'running' | 'success' | 'failed' | 'cancelled'
  page?: number
  page_size?: number
}

export interface TaskCreateReq {
  workflow_id: string
  asset_id?: string
  input_params?: Record<string, any>
}

export interface TaskUpdateReq {
  status?: 'pending' | 'running' | 'success' | 'failed' | 'cancelled'
  progress?: number
  current_node?: string
  error?: string
}

export interface TaskStats {
  total: number
  pending: number
  running: number
  success: number
  failed: number
  cancelled: number
}

export interface TaskContext {
  task_id: string
  version: number
  data?: Record<string, any>
  updated_at: string
}

export interface TaskContextPatch {
  id: string
  task_id: string
  writer_node_key: string
  before_version: number
  after_version: number
  diff: {
    set?: Record<string, any>
    unset?: string[]
  }
  created_at: string
}

export interface TaskContextPatchListResponse {
  items: TaskContextPatch[]
  total: number
}

export interface TaskRunEvent {
  id: string
  task_id: string
  session_id?: string
  event_type: string
  source: string
  node_key?: string
  tool_name?: string
  payload?: Record<string, any>
  created_at: string
}

export interface TaskRunEventListResponse {
  items: TaskRunEvent[]
  total: number
}

export interface TaskListResponse {
  items: Task[]
  total: number
  page: number
  page_size: number
}

export const taskApi = {
  list(params?: TaskListQuery) {
    return apiClient.get<TaskListResponse>('/tasks', { params })
  },

  get(id: string, withRelations: boolean = false) {
    return apiClient.get<Task>(`/tasks/${id}`, { params: { with_relations: withRelations } })
  },

  create(data: TaskCreateReq) {
    return apiClient.post<Task>('/tasks', data)
  },

  update(id: string, data: TaskUpdateReq) {
    return apiClient.put<Task>(`/tasks/${id}`, data)
  },

  delete(id: string) {
    return apiClient.delete(`/tasks/${id}`)
  },

  start(id: string) {
    return apiClient.post<Task>(`/tasks/${id}/start`)
  },

  complete(id: string) {
    return apiClient.post<Task>(`/tasks/${id}/complete`)
  },

  fail(id: string, error: string) {
    return apiClient.post<Task>(`/tasks/${id}/fail`, { error })
  },

  cancel(id: string) {
    return apiClient.post<Task>(`/tasks/${id}/cancel`)
  },

  getStats() {
    return apiClient.get<TaskStats>('/tasks/stats')
  },

  getContext(id: string) {
    return apiClient.get<TaskContext>(`/tasks/${id}/context`)
  },

  listContextPatches(id: string, params?: { limit?: number; offset?: number }) {
    return apiClient.get<TaskContextPatchListResponse>(`/tasks/${id}/context/patches`, { params })
  },

  createContextSnapshot(id: string, trigger: string = 'manual') {
    return apiClient.post(`/tasks/${id}/context/snapshot`, { trigger })
  },

  listEvents(id: string, params?: { source?: string; node_key?: string; limit?: number; offset?: number }) {
    return apiClient.get<TaskRunEventListResponse>(`/tasks/${id}/events`, { params })
  }
}

File: internal/domain/workflow/task.go
package workflow

import (
	"time"

	"github.com/google/uuid"
)

type TaskStatus string

const (
	TaskStatusPending   TaskStatus = "pending"
	TaskStatusRunning   TaskStatus = "running"
	TaskStatusSuccess   TaskStatus = "success"
	TaskStatusFailed    TaskStatus = "failed"
	TaskStatusCancelled TaskStatus = "cancelled"
)

type NodeExecutionStatus string

const (
	NodeExecPending NodeExecutionStatus = "pending"
	NodeExecRunning NodeExecutionStatus = "running"
	NodeExecSuccess NodeExecutionStatus = "success"
	NodeExecFailed  NodeExecutionStatus = "failed"
	NodeExecSkipped NodeExecutionStatus = "skipped"
)

type NodeExecution struct {
	NodeKey     string              `json:"node_key"`
	Status      NodeExecutionStatus `json:"status"`
	Error       string              `json:"error,omitempty"`
	StartedAt   *time.Time          `json:"started_at,omitempty"`
	CompletedAt *time.Time          `json:"completed_at,omitempty"`
	ArtifactIDs []uuid.UUID         `json:"artifact_ids,omitempty"`
}

type Task struct {
	ID                 uuid.UUID
	TenantID           uuid.UUID
	TriggeredByUserID  *uuid.UUID
	WorkflowID         uuid.UUID
	WorkflowRevisionID *uuid.UUID
	WorkflowRevision   int64
	AssetID            *uuid.UUID
	Status             TaskStatus
	Progress           int
	CurrentNode        string
	ContextVersion     int64
	InputParams        map[string]interface{}
	Error              string
	NodeExecutions     []NodeExecution
	StartedAt          *time.Time
	CompletedAt        *time.Time
	CreatedAt          time.Time
	UpdatedAt          time.Time
}

func (t *Task) IsPending() bool {
	return t.Status == TaskStatusPending
}

func (t *Task) IsRunning() bool {
	return t.Status == TaskStatusRunning
}

func (t *Task) IsSuccess() bool {
	return t.Status == TaskStatusSuccess
}

func (t *Task) IsFailed() bool {
	return t.Status == TaskStatusFailed
}

func (t *Task) IsCancelled() bool {
	return t.Status == TaskStatusCancelled
}

func (t *Task) IsCompleted() bool {
	return t.Status == TaskStatusSuccess || t.Status == TaskStatusFailed || t.Status == TaskStatusCancelled
}

func (t *Task) Start() {
	now := time.Now()
	t.Status = TaskStatusRunning
	t.StartedAt = &now
}

func (t *Task) Complete() {
	now := time.Now()
	t.Status = TaskStatusSuccess
	t.Progress = 100
	t.CompletedAt = &now
}

func (t *Task) Fail(errMsg string) {
	now := time.Now()
	t.Status = TaskStatusFailed
	t.Error = errMsg
	t.CompletedAt = &now
}

func (t *Task) Cancel() {
	now := time.Now()
	t.Status = TaskStatusCancelled
	t.CompletedAt = &now
}

func (t *Task) Duration() float64 {
	if t.StartedAt == nil {
		return 0
	}
	endTime := time.Now()
	if t.CompletedAt != nil {
		endTime = *t.CompletedAt
	}
	return endTime.Sub(*t.StartedAt).Seconds()
}

type TaskFilter struct {
	WorkflowID        *uuid.UUID
	AssetID           *uuid.UUID
	TriggeredByUserID *uuid.UUID
	Status            *TaskStatus
	From              *time.Time
	To                *time.Time
	Limit             int
	Offset            int
}

type TaskStats struct {
	Total     int64
	Pending   int64
	Running   int64
	Success   int64
	Failed    int64
	Cancelled int64
}
